api/client/webclient/webclient.go

```diff
@@ -234,3 +255,89 @@ type GithubSettings struct {
 	// Display is the connector display name
 	Display string `json:"display"`
 }
+
+// The tunnel addr is retrieved in the following preference order:
+//  1. Reverse Tunnel Public Address.
+//  2. SSH Proxy Public Address Host + Tunnel Port.
+//  3. HTTP Proxy Public Address Host + Tunnel Port.
+//  4. Tunnel Listen Address.
+func tunnelAddr(settings SSHProxySettings) (string, error) {
+	// If a tunnel public address is set, nothing else has to be done, return it.
+	if settings.TunnelPublicAddr != "" {
+		return extractHostPort(settings.TunnelPublicAddr)
+	}
+
+	// Extract the port the tunnel server is listening on.
+	tunnelPort := strconv.Itoa(defaults.SSHProxyTunnelListenPort)
+	if settings.TunnelListenAddr != "" {
+		if port, err := extractPort(settings.TunnelListenAddr); err == nil {
+			tunnelPort = port
+		}
+	}
+
+	// If a tunnel public address has not been set, but a related HTTP or SSH
+	// public address has been set, extract the hostname but use the port from
+	// the tunnel listen address.
+	if settings.SSHPublicAddr != "" {
+		if host, err := extractHost(settings.SSHPublicAddr); err == nil {
+			return net.JoinHostPort(host, tunnelPort), nil
+		}
+	}
+	if settings.PublicAddr != "" {
+		if host, err := extractHost(settings.PublicAddr); err == nil {
+			return net.JoinHostPort(host, tunnelPort), nil
+		}
+	}
+
+	// If nothing is set, fallback to the tunnel listen address.
+	return extractHostPort(settings.TunnelListenAddr)
```

Comment:
I don't think you want to fallback like this, because `tunnel_listen_addr` might be `0.0.0.0:3024`. Looking at the original code, `tctl` seems to have a more sensible fallback where it would try original address that was dialed, why make this change, what do you think?

```
{
  "auth": {
    "type": "local",
    "second_factor": "off",
    "u2f": {
      "app_id": ""
    },
    "has_motd": false
  },
  "proxy": {
    "kube": {},
    "ssh": {
      "listen_addr": "[::]:3023",
      "tunnel_listen_addr": "0.0.0.0:3024",
      "public_addr": "proxy.example.com:3080",
      "ssh_tunnel_public_addr": "localhost:3024"
    },
    "db": {}
  },
  "server_version": "7.0.0-beta.1",
  "min_client_version": "6.0.0"
}
```

Reply:
Yeah that's definitely more sensible, updated.